### ISSUE TITLE
SEP 037 -- Component References

### ISSUE THREAD

**Issue by jakebeal** [SynBioDex/SEPs#82]
This SEP proposes a ComponentReference object that allows a ComponentInstance to be referenced through children, rather than requiring it to be imported as an immediate child of the containing ComponentDefinition or ModuleDefinition. A new ComponentEquivalence class then takes advantage of this to eliminate the need for MapsTo when linking equivalent components.

https://github.com/SynBioDex/SEPs/blob/master/sep_037.md

Note that this is a competing SEP with the independently developed [SEP 034](https://github.com/SynBioDex/SEPs/issues/79)

**Comment by cjmyers** [comment]
This looks promising, but I disagree on one point.  Namely, I believe useLocal and useRemote are actually very important and they need to be maintained.  Even in your example where you are replacing a generic promoter with a specific one we have to use this.  Since each Component has a definition and they are different, which one do we take?  This is how the useLocal/useRemote decides this.  

Somewhat ironically, you are re-inventing the SBML Replacement/ReplacedBy constructs.  When we introduced MapsTo, I actually made a pitch to swap that into SBML instead, since it has some nice properties, where Replacement/ReplacedBy can be awkward.  Please take a look here at how they do the same thing in SBML:

http://co.mbine.org/specifications/sbml.level-3.version-1.comp.version-1.release-3.pdf

The key classes are ReplacedElement, ReplacedBy, and SBaseRef.  The ReplacedElement is essentially, the useLocal, while the ReplacedBy is essentially the useRemote.  The SBaseRef allows for recursive references just like your ComponentReference class.  In iBioSim, we found these nested references tedious, so we actually back fill in Ports and direct references to Ports to avoid these.  

I see your motivation for supporting these recursive references, but we also had some issues with them.  One of my main sticky points was that it does not promote modular design principles.  If you can refer to anything in any layer of your design, then you don't think about how a module connects into a system.  This would be like allowing in software a connection to local variables several layers deep in function calls.  Granted, biology is different, and there are indeed some headaches when you only discover later something should have been pulled out to a port.  Also, it does mean that you end up with lots of local copies of variables, since you need to instantiate an instance to connect through.

In summary, while I'm sympathetic to the goals here, and I do fight with MapsTos a lot myself, I would like to see first how merging CD and MD affects these issues.  I think that a lot of the MapsTos headaches do go away with that.  The remaining uses of MapsTos will include those used for making connections of inputs and outputs, which I believe MapsTos does well.  To see if there are still other problematic ones, it would be good to redo your examples with CD/MD merged.  

One last thing, I think that auto-generated MapsTos by VPR, etc. are fine as long as there is a flatten routine for visualization and analysis.  This is currently prevented by the CD/MD split.  So, I think that merge is also the key issue there as well.

**Comment by jakebeal** [comment]
On your points:
1. Can you show any real "in the wild" examples with useLocal or useRemote (vs. the others)?  My impression is that in most cases the elements can be effectively merged.
2. It's a massive problem if you can't make a reference without modifying a published component. 
3. I'd be fine with recommending that one not reference anything that's not "exposed" via ports, but the fact of the matter is that we will need to express unintended interactions that do not respect modular design principles.

**Comment by cjmyers** [comment]
Your third example fails for merge.  It has two components which point to different CDs via their definition link.  A Component is only allowed a single definition, so there is no way to merge.  You need to know which definition to keep when you flatten.  In fact, merge has never been defined formally for flatten and likely cannot be, since we have other singleton fields like name and description, which would you keep in a merge?  You cannot keep both.

Yes, your point (2) is the reason that SBML used to motivate the recursive SBaseRefs.  Those of us that preferred Ports in the SBML community felt that if you want to connect to a model in a new way, you should copy and add the ports you need.  Publishing a new version with new "ports" seems to be a reasonable solution in this case.  It is what you would do in software (or hardware), if you suddenly wanted access to an internal variable (or state).

Your point (3) is certainly fair, and it was the point the also won the day for recursive SBaseRef in SBML.  Biology just is not as clean in its interfaces, so having this capability is likely going to be needed even if not encouraged.  

Just talked with SBOL editors about these new SEPs, and this and SEP 035 are essentially trying to add SBOL3 functionality to SBOL2.  The danger in both of these SEPs is that they could lead to a delay in SBOL3.  They also make for a pretty complicated migration.  Namely, we will likely need converters to support these.  It would be cleaner if we just added these new functionalities into SBOL3 and try to get that done sooner than later.  We will have converters for SBOL2 to/from SBOL3, but I was hoping to avoid them in SBOL2.x.  Having two ways to represent the same thing in the same version is going to potentially cause some significant challenges for exchange without these converters.

**Comment by jakebeal** [comment]
> Your third example fails for merge.

I don't believe that it fails, I believe that it's a reasonable model to express. I believe this instead is a good example of why we need to define a correct behavior for flattening.

**Comment by cjmyers** [comment]
How are you proposing to know which is the template and which is the real design?  If you say based on existence of sequence or something similar, I don’t like it.  It requires too much special casing.  The useLocal/useRemote is much cleaner, general, and explicit.  If you try to flatten your 3rd example, you should see that there is no way to do it, if you use refinement merge.

**Comment by jakebeal** [comment]
I'm proposing that every attribute of template should be either blank (e.g., name, sequence) or else compatible with the instantiation (e.g., type and role).  And if they aren't, then that's a merge exception that requires human intervention to sort out.

**Comment by cjmyers** [comment]
I really don’t think this would work.  There are required attributes, such as definition for the Component that cannot be blank or compatible.  I also don’t think it is good to then propagate this down to the CDs.  

In hierarchical design, you want to be able to replace out things on ports with not only more specific things, but genuine replacements.  You might want to say that you want to use a particular gate, but with the RBS replaced with a different one.  These types of replacements would not be covered by your approach.  You might want to replace an engineered region with a specified part.  

I’m not very supportive of adding new features even in SBOL3 that make some features in SBOL2 unable to be converted.  I think we should always have a path for up-conversion.

**Comment by jakebeal** [comment]
Do you have any real-life examples of where we've _actually_ done such a replacement?

**Comment by cjmyers** [comment]
We heavily use MapsTo in our model generator.  There are some examples on LCP SBH, but it is down at the moment (certificate is being updated).  The CRISPR tutorial example uses this form of MapsTo.  If I had made this proposal, I would have eliminated the ill-defined merge, and left the useLocal and useRemote.  They are much cleaner and follow well understand paradigms used in both hardware and software languages.  The merge was never worked out, and in principle, I don't think it can be except in extremely limited and restricted ways.  I think your proposal of recursive references is fine, but you just need to add some mechanism to indicate how the replacement is to be made.  

I think you can draw inspiration from the SBML comp package that was debated and worked on for years before it was accepted.  It has some quirks (which I think your proposal shares) in the accessing of these relationships, but it can be much more compact like in your case.  The key idea is the notion of a replacement.  

In your last example, you would change ComponentEquivalence to a ComponentReplacement, it would then have one property which is the thing doing the replacing, and another which is a list of the things being replaced.  For verifyIdentical or merge, you would not be required to provide the item doing the replacing, but you would need to indicate if this is verifyIdentical or merge.

One more thing this proposal is lacking that SBML comp has is the notion of deletion.  What if you want to include something without something else?  This could also be handled with my scheme, you could also have an empty replacement object, but in this case, you would indicate this is a replacement, so it would replace the items with nothing.

While I understand your desire for use cases, keep in mind that outside of SynBio, we have tons of examples of hierarchical composition and these place certain requirements on functionality.  If we cannot currently support these easily (which is true), then the examples are not there.   However, if we have a nice general scheme as this could develop to, then we will likely see these use cases.

**Comment by graik** [comment]
As you probably guess, I am very sympathetic with the idea of getting rid of or reducing the use of `mapsTo` :) However, this proposal still seems unnecessarily complex. Do we really need the concept of *replacement* and *equivalence*? If I understand correctly, this still follows the philosophy that we can/must define new functional devices by replacing sub-parts in other devices. Just imagine a programming language would allow you to import a library but replace line 112 to 148 before using it. It would be a complete nightmare which is why this flexibility is not allowed.

Since we are about to merge ModuleDefinition and ComponentDefinition and remove FunctionalComponent, we can finally utilize the very well designed `Interaction` for a clean (mostly hierarchical) description of functional modules. It's all there already. We only need to allow to have Participation point to ComponentDefinition or to another Interaction for hierarchical composition.

Example: Definition of a transcription unit:

**Interaction** GFP-Production:
   * type: ProteinGenerator 
   * **Participation**:
       * role: promoter
       * participant: -> Interaction(type=transcriptionActivation) -> promoter+transcriptionfactor
   * **Participation**:
       * role: CDS
       * participant: -> ComponentDefintion (GFP gene)
   * **Participation**:
       * role: terminator
       * participant: -> ComponentDefintion (terminator)
   * **Participation**:
       * role: product
       * participant: -> ComponentDefinition (GFP protein)

Abstract parts can be easily defined by declaring Participations with role but without participants.

**Comment by bbartley** [comment]
> Just imagine a programming language would allow you to import a library but replace line 112 to 148 before using it. It would be a complete nightmare which is why this flexibility is not allowed.

Actually, something like this is allowed in Python... it is called "monkey-patching".  Sometimes it's very handy, and, yes, sometimes it is a complete nightmare...

**Comment by graik** [comment]
> > Just imagine a programming language would allow you to import a library but replace line 112 to 148 before using it. It would be a complete nightmare which is why this flexibility is not allowed.
> 
> Actually, something like this is allowed in Python... it is called "monkey-patching". Sometimes it's very handy, and, yes, sometimes it is a complete nightmare...

Ok, I saw this coming 😛... It is true that Python programs can, in theory, completely rewrite themselves at runtime. Switching  the type / class of an already existing object is the most I have ever done in this respect. Even  in the worst monkey patching, I have never seen anyone replacing actual code segments...

**Comment by jakebeal** [comment]
@graik I believe you're responding to Chris' suggestions rather than my actual proposal. In my proposal, there is no replacement, only identity and merge.

The rest of your discussion, regarding interactions, is interesting but I believe is aiming to solve a different problem than what this SEP is targeted at.

**Comment by jakebeal** [comment]
I've updated this SEP based on the discussion in the SBOL 3 working group.  The new version has:
- ComponentReference and ComponentInstance become sibling classes under SubComponent.
- a generalization of ComponentEquivalence to DesignConstraint, also addressing one of our pending generalizations of SequenceConstraint.  I am keeping SequenceConstraint pending resolution of SEP 025, however.

**Comment by cjmyers** [comment]
I think we may want to set an order of these SEPs where we approve SEP 025 first.  The UML diagram should be much simpler if that is approved.  I would prefer the parent class to be ComponentInstance (or SubComponentInstance), then the children would be Component and ComponentReference (or SubComponent and SubComponentReference).  We definitely need to move everything up to the Parent class, since these can certainly have Direction, Roles, RoleIntegrations, a sourceLocations, and eventually Locations too.  

Also for DesignConstraint, I thought you wanted multiple Objects, so you could say one thing replaces many others.  We might consider extending subject and object into lists.  Just a thought.  Will work either way.

**Comment by jakebeal** [comment]
When SEP 025 settles, I plan to update this to reflect its state of consensus.

**Comment by jakebeal** [comment]
I've updated this with my understanding of the likely SBOL 3 class hierarchy, shared with SEP 037 and SEP 043.

Per discussion, I also changed DesignConstraint to just be a renaming of SequenceConstraint and addition of new recommended values.

**Comment by cjmyers** [comment]
Again, I'm not excited about making SubComponent an abstract class, since this will mean SubComponent never appears in serialization.  Since SEP 043 is proposing making definition optional, then I would prefer to merge SubComponent and ComponentInstance in the UML with an optional definition attribute.

**Comment by graik** [comment]
Sounds good to me! Simple is better :) and "empty" SubComponents with an assigned role are perfect to indicate let software suggest possible parts that can fill that gap. We are also discussing this as a templating mechanism

**Comment by jakebeal** [comment]
@cjmyers Other than the SubComponent naming issue, are you OK with the proposed ComponentReference class?  I believe that it should _not_ be merged with the others, since it's got significantly different fields.

**Comment by cjmyers** [comment]
If you flip the names, I can get behind this proposal and the others.  I'm also ok with merging which make things simpler, even if it means having some fields used in some place and not in others.

**Comment by jakebeal** [comment]
Per discussion today, I have changed "DesignConstraint" to "Constraint", as we will also often want to use this to describe constraints that weren't necessarily "designed" per se
